components/authz-service/server/v2/system_integration_test.go

```diff
@@ -118,91 +144,314 @@ func TestListRulesForAllProjects(t *testing.T) {
-		})
-		require.NoError(t, err)
-		createResp2, err := ts.projects.CreateRule(ctx, &api_v2.CreateRuleReq{
-			Id:         id2,
-			Name:       name,
-			Type:       type2,
-			ProjectId:  pid1,
-			Conditions: conditions2,
-		})
-		require.NoError(t, err)
-		createResp3, err := ts.projects.CreateRule(ctx, &api_v2.CreateRuleReq{
-			Id:         id3,
-			Name:       name,
-			Type:       type2,
-			ProjectId:  pid2,
-			Conditions: conditions3,
-		})
-		require.NoError(t, err)
-
-		expectedRules1 := []*api_v2.ProjectRule{createResp1.Rule, createResp2.Rule}
-		expectedRules2 := []*api_v2.ProjectRule{createResp3.Rule}
-
-		list, err := ts.projects.ListRules(ctx, &api_v2.ListRulesReq{})
-		require.Equal(t, 3, len(list.Rules))
-		require.NoError(t, err)
-
-		resp, err := ts.projects.ListRulesForAllProjects(ctx, &api_v2.ListRulesForAllProjectsReq{})
-		require.NoError(t, err)
-		actualPid1Rules, ok := resp.ProjectRules[pid1]
-		require.True(t, ok)
-		actualPid2Rules, ok := resp.ProjectRules[pid2]
-		require.True(t, ok)
-		assert.ElementsMatch(t, expectedRules1, actualPid1Rules.Rules)
-		assert.ElementsMatch(t, expectedRules2, actualPid2Rules.Rules)
+			_, err = ts.Projects.UpdateRule(ctx, &api_v2.UpdateRuleReq{
+				Id:         createResp2.Rule.Id,
+				ProjectId:  createResp2.Rule.ProjectId,
+				Name:       "this has been updated",
+				Type:       createResp2.Rule.Type,
+				Conditions: newConditions,
+			})
+			require.NoError(t, err)
+
+			// Re-apply and check cache
+			_, err = ts.Projects.ApplyRulesStart(ctx, &api_v2.ApplyRulesStartReq{})
+			assert.NoError(t, err)
+
+			expectedAfterApplyRules1 := []*api_v2.ProjectRule{
+				&api_v2.ProjectRule{
+					Id:         createResp2.Rule.Id,
+					ProjectId:  createResp2.Rule.ProjectId,
+					Name:       "this has been updated",
+					Type:       createResp2.Rule.Type,
+					Conditions: newConditions,
+					Status:     "applied",
+				},
+			}
+			eexpectedAfterApplyRules2 := []*api_v2.ProjectRule{createResp3.Rule}
+
+			resp2, err := ts.Projects.ListRulesForAllProjects(ctx, &api_v2.ListRulesForAllProjectsReq{})
+			require.NoError(t, err)
+			actualAfterApplyPid1Rules, ok := resp2.ProjectRules[pid1]
+			require.True(t, ok)
+			actualAfterApplyPid2Rules, ok := resp2.ProjectRules[pid2]
+			require.True(t, ok)
+			rulesEqual(t, expectedAfterApplyRules1, actualAfterApplyPid1Rules.Rules)
+			rulesEqual(t, eexpectedAfterApplyRules2, actualAfterApplyPid2Rules.Rules)
+
+			ts.Shutdown(t, ctx)
+		}},
+	}
+
+	rand.Shuffle(len(cases), func(i, j int) {
+		cases[i], cases[j] = cases[j], cases[i]
 	})
+
+	for _, test := range cases {
+		t.Run(test.desc, test.f)
+	}
 }
 
-func setupWithOPAV2(t *testing.T) testSetup {
+func rulesEqual(t *testing.T, expected, actual []*api_v2.ProjectRule) {
+	t.Helper()
+	if len(expected) != len(actual) {
+		t.FailNow()
+		fmt.Printf("expected len %d != actual len %d", len(expected), len(actual))
+	}
+
+	for i, expectedRule := range expected {
+		actualRule := actual[i]
```

Comment:
While it might work coincidentally, relying on the DB to return rules in the same order makes the tests fragile. Consider turning one into a hash on the id so you could then iterate the other and find the right rule, then compare them.